spec.md

```diff
@@ -0,0 +1,1679 @@
+ |              +---+----+---+              |
+ |       Controller |    | Controller       v
++++         Publish |    | Unpublish       +++
+|X|          Volume |    | Volume          | |
++-+             +---v----+---+             +-+
+                | NODE_READY |
+                +---+----^---+
+               Node |    | Node
+            Publish |    | Unpublish
+             Volume |    | Volume
+                +---v----+---+
+                | PUBLISHED  |
+                +------------+
+
+Figure 4: The lifecycle of a dynamically provisioned volume, from
+creation to destruction.
+```
+
+```
+    Controller                  Controller
+       Publish                  Unpublish
+        Volume  +------------+  Volume
+ +------------->+ NODE_READY +--------------+
+ |              +---+----^---+              |
+ |             Node |    | Node             v
++++         Publish |    | Unpublish       +++
+|X| <-+      Volume |    | Volume          | |
++++   |         +---v----+---+             +-+
+ |    |         | PUBLISHED  |
+ |    |         +------------+
+ +----+
+   Validate
+   Volume
+   Capabilities
+
+Figure 5: The lifecycle of a pre-provisioned volume that requires
+controller to publish to a node (`ControllerPublishVolume`) prior to
+publishing on the node (`NodePublishVolume`).
+```
+
+```
+       +-+  +-+
+       |X|  | |
+       +++  +^+
+        |    |
+   Node |    | Node
+Publish |    | Unpublish
+ Volume |    | Volume
+    +---v----+---+
+    | PUBLISHED  |
+    +------------+
+
+Figure 6: Plugins may forego other lifecycle steps by contraindicating
+them via the capabilities API. Interactions with the volumes of such
+plugins is reduced to `NodePublishVolume` and `NodeUnpublishVolume`
+calls.
+```
+
+The above diagrams illustrate a general expectation with respect to how a CO MAY manage the lifecycle of a volume via the API presented in this specification.
+Plugins MAY or MAY NOT expose RPCs to facilitate `CreateVolume`/`DestroyVolume` or `AttachVolume`/`DetachVolume` operations.
+The full list of plugin capabilities is documented in the `GetControllerServiceCapabilities` and `GetNodeServiceCapabilities` RPCs.
+
+## Container Storage Interface
+
+This section describes the interface between COs and Plugins.
+
+### RPC Interface
+
+A CO interacts with an Plugin through RPCs. Each SP MUST provide:
+
+* **Node Plugin**: A gRPC endpoint serving CSI RPCs that MUST be run on the Node whereupon an SP-provisioned volume will be published.
+* **Controller Plugin**: A gRPC endpoint serving CSI RPCs that MAY be run anywhere.
+* In some circumstances a single gRPC endpoint MAY serve all CSI RPCs (see Figure 3 in Architecture).
+
+There are three sets of RPCs:
+
+* **Identity Service**: Both the Node Plugin and the Controller Plugin MUST implement this sets of RPCs.
+* **Controller Service**: The Controller Plugin MUST implement this sets of RPCs.
+* **Node Service**: The Node Plugin MUST implement this sets of RPCs.
+
+```proto
```

Comment:
nit: the remaining protobuf sections are typed `protobuf` instead of `proto`, we should probably be consistent this would make it easier, for example, to write a sed/awk script that extracts all the protobuf sections from this document (for passing to a protoc compiler)